File: 42. Trapping Rain Water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        max1 = 0
        max2 = 0
        left = [0]*len(height)
        right = [0]*len(height)
        j = len(height)-1
        total_area = 0
        for i in range(len(height)):
            left[i] = max1
            right[j] = max2
            max1 = max(max1,height[i])
            max2 = max(max2,height[j])
            j -= 1
        for x in range(len(left)):
            min1 = min(left[x],right[x])
            area = min1 - height[x]
            if area < 0:
                area = 0
            total_area += area
        return total_area      

File: 42. Trapping Rain Water/solution.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut max1 = 0;
        let mut max2 = 0;
        let n = height.len();
        let mut j = n-1;
        let mut left = vec![0;n];
        let mut right = vec![0;n];
        for i in 0..n{
            left[i] = max1;
            right[j] = max2;
            max1 = std::cmp::max(max1,height[i]);
            max2 = std::cmp::max(max2,height[j]);
            j -= 1;
        }
        let mut max_area = 0;
        for x in 0..n{
        let area = std::cmp::min(left[x],right[x]) - height[x];
        if area > 0{
            max_area += area;
        }   
        } max_area
    }
}